infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,293 @@
+     * @return A list of addresses, currently not present in the stream log.
+     */
+    ImmutableList<Long> getUnknownAddressesInRange(long rangeStart, long rangeEnd) {
+        Set<Long> knownAddresses = streamLog
+                .getKnownAddressesInRange(rangeStart, rangeEnd);
+
+        return LongStream.range(rangeStart, rangeEnd + 1L)
+                .filter(address -> !knownAddresses.contains(address))
+                .boxed()
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Performs the state transfer for the current transfer segments and also
+     * updates their state as a result.
+     *
+     * @param transferSegments A list of the segment currently present in the system.
+     * @return A list with the updated transfer segments.
+     */
+    public ImmutableList<TransferSegment> handleTransfer(List<TransferSegment> transferSegments) {
+
+        return transferSegments.stream().map(segment -> {   // For each of the segments:
+                    TransferSegmentStatus newStatus = segment.getStatus();
+
+                    // If a segment is not NOT_TRANSFERRED -> return it as is.
+                    if (newStatus.getSegmentState() != NOT_TRANSFERRED) {
+                        return segment;
+                    }
+
+                    // Get all the unknown addresses for this segment.
+                    List<Long> unknownAddressesInRange =
+                            getUnknownAddressesInRange(segment.getStartAddress(),
+                                    segment.getEndAddress());
+
+                    // If no addresses to transfer - mark a segment as transferred.
+                    if (unknownAddressesInRange.isEmpty()) {
+                        log.debug("All addresses are present in a range, skipping transfer.");
+                        long totalTransferred = segment.computeTotalTransferred();
+
+                        newStatus = TransferSegmentStatus
+                                .builder()
+                                .segmentState(TRANSFERRED)
+                                .totalTransferred(totalTransferred)
+                                .build();
+                    } else {
+                        // Get total number of addresses needed to transfer.
+                        long numAddressesToTransfer = unknownAddressesInRange.size();
+                        // Create a transferBatchRequest stream.
+                        Stream<TransferBatchRequest> batchStream = Lists
+                                .partition(unknownAddressesInRange, batchSize)
+                                .stream()
+                                .map(groupedAddresses ->
+                                        new TransferBatchRequest(groupedAddresses, Optional.empty())
+                                );
+                        // Execute state transfer synchronously.
+                        newStatus = synchronousStateTransfer(batchStream, numAddressesToTransfer);
```

Comment:
I thought one of your goals was to add parallelism. This will be executed on the same workflow thread.

Reply:
When I first introduced the PR it had these changes, then we decided to cut out all the S/T parts, and introduce S/T within a few PRs. The main reason was that PR was too big for review. 
This is the first part of the State Transfer Redesign. 
The main goal of this PR is to make state transfer execute locally as per the design doc. 
This PR is mostly concerned with changing the clustering model and making the changes for S/T that would work with this model.
All the S/T related optimizations are coming to next PRs.